packages/vhd-lib/src/Vhd/VhdDirectory.js

```diff
@@ -0,0 +1,186 @@
+    start /= SECTOR_SIZE // in sector
+    const blockSectorSize = this.fullBlockSize / SECTOR_SIZE
+    let nbNonEmptyblock = 0
+    for (let blockId = 0; blockId < this.header.maxTableEntries; blockId++) {
+      if (test(buffer, blockId)) {
+        this.blockTable.writeUInt32BE(start + nbNonEmptyblock * blockSectorSize, blockId * 4)
+        nbNonEmptyblock++
+      } else {
+        this.blockTable.writeUInt32BE(BLOCK_UNUSED, blockId * 4)
+      }
+    }
+  }
+
+  containsBlock(blockId) {
+    assert.notStrictEqual(this.blockTable, undefined, 'Block table must not be empty to access a block address')
+
+    return this.blockTable.readUInt32BE(blockId * 4) !== BLOCK_UNUSED
+  }
+
+  getChunkPath(partName) {
+    return this._path + '/' + partName
+  }
+
+  async _readChunk(partName) {
+    // here we can implement compression and / or crypto
+    const buffer = await this._handler.readFile(this.getChunkPath(partName))
+
+    return {
+      buffer: Buffer.from(buffer),
+    }
+  }
+
+  async _writeChunk(partName, buffer) {
+    assert(Buffer.isBuffer(buffer))
+    // here we can implement compression and / or crypto
+    return this._handler.writeFile(this.getChunkPath(partName), buffer, { flags: this._flags })
+  }
+
+  async readHeaderAndFooter() {
+    const { buffer: bufHeader } = await this._readChunk('header')
+    const { buffer: bufFooter } = await this._readChunk('footer')
+    const footer = buildFooter(bufFooter)
+    const header = buildHeader(bufHeader, footer)
+
+    this.footer = footer
+    this.header = header
+  }
+
+  async readBlock(blockId, onlyBitmap = false) {
+    if (onlyBitmap) {
+      throw new Error(`reading 'bitmap of block' ${blockId} in a VhdDIrecotry is not implemented`)
+    }
+    const { buffer } = await this._readChunk(blockId)
+    return {
+      id: blockId,
+      bitmap: buffer.slice(0, this.bitmapSize),
+      data: buffer.slice(this.bitmapSize),
+      buffer,
+    }
+  }
+  ensureBatSize() {
+    // nothing to do in directory mode
+  }
+
+  async writeFooter() {
+    const { footer } = this
+
+    const rawFooter = fuFooter.pack(footer)
+
+    footer.checksum = checksumStruct(rawFooter, fuFooter)
+    debug(`Write footer  (checksum=${footer.checksum}). (data=${rawFooter.toString('hex')})`)
+
+    await this._writeChunk('footer', rawFooter)
+  }
+
+  writeHeader() {
+    const { header } = this
+    const rawHeader = fuHeader.pack(header)
+    header.checksum = checksumStruct(rawHeader, fuHeader)
+    debug(`Write header  (checksum=${header.checksum}). (data=${rawHeader.toString('hex')})`)
+    return this._writeChunk('header', rawHeader)
+  }
+
+  writeBlockAllocationTable() {
+    assert.notStrictEqual(this.blockTable, undefined, 'Block allocation table has not been read')
+    assert(this.blockTable.length, 'Block allocation table is empty')
+    assert(this.blockTable.length % 4 === 0, 'Block allocation table size is incorrect')
+
+    const { blockTable } = this
```

Comment:
Move this at the beginning of the function, it will make the assertions simpler.